flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/ByteBufferReadableFSDataInputStream.java

```diff
@@ -0,0 +1,195 @@
+
+    private final FSDataInputStream originalInputStream;
+
+    /**
+     * InputStream Pool which provides multiple input streams to random read concurrently. An input
+     * stream should only be used by a thread at a point in time.
+     */
+    private final Queue<FSDataInputStream> readInputStreamPool;
+
+    private final Callable<FSDataInputStream> inputStreamBuilder;
+
+    public ByteBufferReadableFSDataInputStream(
+            FSDataInputStream originalInputStream,
+            Callable<FSDataInputStream> inputStreamBuilder,
+            int inputStreamCapacity) {
+        this.originalInputStream = originalInputStream;
+        this.inputStreamBuilder = inputStreamBuilder;
+        this.readInputStreamPool = new LinkedBlockingQueue<>(inputStreamCapacity);
+    }
+
+    /**
+     * Reads up to <code>ByteBuffer#remaining</code> bytes of data from the input stream into a
+     * ByteBuffer. Not Thread-safe yet since the interface of sequential read of ForSt only be
+     * accessed by one thread at a time.
+     *
+     * @param bb the buffer into which the data is read.
+     * @return the total number of bytes read into the buffer.
+     * @exception IOException If the first byte cannot be read for any reason other than end of
+     *     file, or if the input stream has been closed, or if some other I/O error occurs.
+     * @exception NullPointerException If <code>bb</code> is <code>null</code>.
+     */
+    public int readFully(ByteBuffer bb) throws IOException {
+        if (bb == null) {
+            throw new NullPointerException();
+        } else if (bb.remaining() == 0) {
+            return 0;
+        }
+        return readFullyFromFSDataInputStream(originalInputStream, bb);
+    }
+
+    /**
+     * Reads up to <code>ByteBuffer#remaining</code> bytes of data from the specific position of the
+     * input stream into a ByteBuffer. Tread-safe since the interface of random read of ForSt may be
+     * concurrently accessed by multiple threads. TODO: Support to split this method to other class.
+     *
+     * @param position the start offset in input stream at which the data is read.
+     * @param bb the buffer into which the data is read.
+     * @return the total number of bytes read into the buffer.
+     * @exception IOException If the first byte cannot be read for any reason other than end of
+     *     file, or if the input stream has been closed, or if some other I/O error occurs.
+     * @exception NullPointerException If <code>bb</code> is <code>null</code>.
+     */
+    public int readFully(long position, ByteBuffer bb) throws Exception {
+        if (bb == null) {
+            throw new NullPointerException();
+        } else if (bb.remaining() == 0) {
+            return 0;
+        }
+
+        FSDataInputStream fsDataInputStream = readInputStreamPool.poll();
+        if (fsDataInputStream == null) {
+            fsDataInputStream = inputStreamBuilder.call();
```

Comment:
IIUC, there should be control of stream count here?

Reply:
Right. It's controlled by inputStreamCapacity.

Reply:
I'd be fine with no hard limit control. It concurrent read a common case? Otherwise I'm thinking is it better to make `DEFAULT_INPUT_STREAM_CAPACITY` as `1`?

Reply:
I think concurrent read may be a common case which is controlled in ForSt random read.
I'd prefer to remain this currently, maybe we could test it later and give a better default value for it, WDYT?